File: ClassificationMethods.R
##### Classification Methods ####

##### Libraries #####

library(ISLR2)
library(tidyverse)
library(MASS)
library(klaR)
library(e1071)
library(class)
library(caret)
library(ROCR)
library(nortest)
library(MVN)
library(biotools)

##### Functions #####

#Function for evaluate p-values

tStest = function(x){
  
  nom = names(x)
  a = rep(0,length(x))
  
  for(i in seq_along(x)){
    
    if(x[i] > 0.025){a[i]='Accept Ho'}
    
    else{a[i]='Reject Ho'}
    
  }
  
  return(tibble('Variables'=nom,'Contraste'=a,"p-valor"=x))
  
}

#Function for measuring accuracy and error

#data = vector of true classes

#Prediction = predicted classes

ClassAccuracy <- function(data,prediction){
  
  confusion_matrix = table(data,prediction,dnn=c("Real Class", "Prediction"))
  
  #Accuracy
  
  #Class accuracy
  
  accuracy.class<-diag(prop.table(confusion_matrix,1))
  
  #Global accuracy
  
  accuracy.global<-sum(diag(prop.table(confusion_matrix)))
  
  #Error
  
  #Error by class
  
  error.class<-diag(1-prop.table(confusion_matrix,1))
  
  #Global error
  
  error.global<-1-accuracy.global
  
  #Output
  
  out <- list('Confusion Matrix'=confusion_matrix,
              'Class Accuracy'=accuracy.class, 'Global Accuracy'=accuracy.global,
              'Class Error'=error.class,'Global Error'=error.global)
  
  print('Measurement of the accuracy')
  
  cat('\n')
  
  return(out)
  
  
}

##### Data #####

datos <- tibble(Smarket)

summary(datos) #descriptive statistics

is.factor(datos$Direction) #Y must be a factor

#if not: Y <- as.factor(Y)

names(datos) #variables

dim(datos) #dimension

datos %>% select(-Direction) %>% cor() #correlaciones

cor(datos[,-9]) #without pipes %>%

#Graphs for volume

plot(datos$Volume,xlab='time',ylab='Volume')

ggplot(datos) +
  geom_point(mapping = aes(seq_along(Volume),y = Volume,colour=Direction))+
  geom_smooth(mapping = aes(seq_along(Volume),y = Volume),alpha=0.4,se=FALSE)+
  labs(x='Time',y='Volume')

##### Logistic Regression #####

#Estimation

logistic.fit <- glm(
  Direction ~ Lag1 + Lag2 + Lag3 + Lag4 + Lag5 + Volume ,
  family = binomial, data = datos)

logistic.fit$fitted.values #fitted values

logistic.fit$coefficients #coefficients

logistic.fit$residuals #residuals

#Summary 

summary(logistic.fit)

summary(logistic.fit)$coefficients #parameters

summary(logistic.fit)$coefficients[,4] #p-values

tStest(summary(logistic.fit)$coefficients[,4])

#Prediction #Gives posteriori probabilities P(Y=k|X=x)

contrasts(datos$Direction) #code for the dummy

logistic.probs <- predict(logistic.fit, type = 'response') #P(Y=1|X=x)

head(logistic.probs)

logistic.pred <- rep("Down", 1250)

logistic.pred[logistic.probs > 0.5] = "Up" #classifies as up if p>0.5

head(logistic.pred)

is.factor(logistic.pred)

logistic.pred = as.factor(logistic.pred)

datos = cbind(datos,logistic.pred) #union of predictions with data

#Graph

datos.logistic <- tibble('Posterior'=logistic.fit$fitted.values,
                         'Index'=seq_along(datos$Direction))

ggplot(datos.logistic) +
  geom_line(aes(Index,Posterior))+
  geom_smooth(aes(Index,Posterior),color='blue')


#Accuracy and error

ClassAccuracy(datos$Direction,logistic.pred)

#Confusion Matrix

(logistic.conf = table(datos$Direction,
                      logistic.pred,dnn=c("Real Class", "Prediction")))

#Accuracy

#Class accuracy

(accuracy.class.logistic<-diag(prop.table(logistic.conf,1)))

#Global accuracy

(accuracy.global.logistic<-sum(diag(prop.table(logistic.conf))))

mean(logistic.pred == datos$Direction)

#Error

#Error by class

(error.class.logistic<-diag(1-prop.table(logistic.conf,1)))

#Global error

(error.global.logistic<-1-accuracy.global.logistic)

1-mean(logistic.pred == datos$Direction)

#Accuracy and error with caret package

confusionMatrix(logistic.pred ,datos$Direction)

##### Logistic Regresion with training/test data ####

#Separating data

train <- (datos$Year < 2005) #data for year < 2005 #boolean

(datos.test <- datos[!train, ]) #test data #without train data

datos.test <- datos %>% filter(Year == 2005) #alternative 

dim(datos.test)

Direction.test <- datos$Direction[!train]

#sample(datos,100) random sample of size n=100

#n = length(datos$Direction) # O dim(datos)[1]
#train = sample(1:n, 0.70*n) #70% for train data

#Fitting the model

logistic.fit.train <- glm(Direction ~ Lag1 + Lag2 + Lag3 + Lag4 + Lag5 + Volume,
                          data = datos , family = binomial , subset = train)

summary(logistic.fit.train)


#Predictions on test data #posteriori probabilities P(Y=1|X=x)

logistic.fit.test.probs <- predict(logistic.fit.train , datos.test,
                                    type = "response")

#Class prediction

logistic.fit.test.pred <- rep("Down", 252)

logistic.fit.test.pred[logistic.fit.test.probs > 0.5] <- "Up" #classify as 1 if p>0.5


#Accuracy on test data

ClassAccuracy(Direction.test, logistic.fit.test.pred)

(logistic.conf.test <- table(Direction.test, logistic.fit.test.pred,
                             dnn=c("Real Class", "Prediction")))

#Class accuracy

(accuracy.class.logistic.test<-diag(prop.table(logistic.conf.test,1)))

#Global accuracy

(accuracy.global.logistic.test<-sum(diag(prop.table(logistic.conf.test))))

mean(logistic.fit.test.pred == datos.test$Direction)

#Error

#Error by class

(error.class.logistic.test<-diag(1-prop.table(logistic.conf.test,1)))

#Global error

(error.global.logistic.test<-1-accuracy.global.logistic.test)

1-mean(logistic.fit.test.pred == datos.test$Direction)

#Accuracy and error with caret package

confusionMatrix(as.factor(logistic.fit.test.pred) ,Direction.test)

##### Linear Discriminant Analysis ####

#Fitting the model 
#With training data

(lda.fit <- lda(Direction ~ Lag1 + Lag2 , data = datos , subset = train))

summary(lda.fit)

#Graphs

plot(lda.fit)

plot(datos[,c("Lag1", "Lag2")], 
     main="Clasification", pch=21,
     bg=c("red", "green")[unclass(datos$Direction)])
abline(lda.fit,col='blue')

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2, method = "lda")

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2,
         method = "lda", plot.matrix = TRUE, imageplot = FALSE)

#Prediction on test data

lda.pred <- predict(lda.fit , datos.test)

names(lda.pred)

lda.pred$class #predicted classes

lda.pred$posterior #posteriori probabilities P(Y=k|X=x)

lda.pred$x #linear discriminant puntuation. delta

#Accuracy and error

(lda.conf <- table(Direction.test, lda.pred$class,
                             dnn=c("Real Class", "Prediction")))

ClassAccuracy(Direction.test, lda.pred$class)

#Accuracy on test data

#Class accuracy

(accuracy.class.lda<-diag(prop.table(lda.conf,1)))

#Global accuracy

(accuracy.global.lda<-sum(diag(prop.table(lda.conf))))

mean(lda.pred$class == datos.test$Direction)

#Error

#Error by class

(error.class.lda<-diag(1-prop.table(lda.conf,1)))

#Global error

(error.global.lda<-1-accuracy.global.lda)

1-mean(lda.pred$class == datos.test$Direction)

#Accuracy and error with caret package

confusionMatrix(lda.pred$class ,Direction.test)

#ROC curve

pred <- prediction(lda.pred$posterior[,1], Direction.test) 

perf <- performance(pred,"tpr","fpr")

plot(perf,colorize=TRUE)

#Lambda de Wilks

X<-as.matrix(datos[,c(1:8)])#Only quantitatives

datos.manova<-manova(X~datos$Direction)

(datos.wilks<-summary(datos.manova, test="Wilks"))

#With eigen values

datos.wilks$Eigenvalues 	# Eigen Values

(lambda<-1/(1+datos.wilks$Eigenvalues[1])) 

#Canonical Correlation. eta^2

(eta2<- datos.wilks$Eigenvalues[1]/(1+ datos.wilks$Eigenvalues[1]))

(corr.canonica<-sqrt(eta2))

#LDA classifies on 1 if P(Y=k|X=x)>0.5. 1=Down

sum(lda.pred$posterior[, 1] >= 0.5)

sum(lda.pred$posterior[, 1] < .5)

#posterior probability threshold

#Change to 0.4

lda.pred$posterior[, 1] > 0.4

lda.posterior = lda.pred$posterior[, 1] 

lda.pred.modified = rep(0,length(lda.posterior))

for(i in seq_along(lda.posterior)){
  
  if(lda.posterior[i] > 0.4){lda.pred.modified[i]='Down'}else{
    lda.pred.modified[i]='Up'}
  
}

lda.pred.modified

#Accuracy

ClassAccuracy(Direction.test, lda.pred.modified)

#Variable selection

#With 10-folders cross-validation rate of correctness

(lda.step <- stepclass(Direction ~ ., "lda", direction = "both", data = datos))

lda.step$formula #best predictors

(lda.best <- lda(lda.step$formula , data = datos , subset = train))

#With Wilk's Lambda

#By Wilk's Lambda

wilks.selection <- greedy.wilks(Direction ~ ., data=datos, niveau=0.1)

wilks.selection$formula #best predictors

(lda.best.wilks <- lda(wilks.selection$formula , data = datos , subset = train))

#LDA with Leave One Out Cross Validation

(lda.fit.cv <- lda(Direction ~ Lag1 + Lag2 , data = datos ,
                   subset = train, CV=TRUE))

##### Quadratic Discriminant Analysis #####

#Fitting the model

(qda.fit <- qda(Direction ~ Lag1 + Lag2 , data = datos ,
               subset = train))

#Graphs

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2, method = "qda")

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2,
         method = "qda", plot.matrix = TRUE, imageplot = FALSE)

#Prediction on test data

qda.class <- predict(qda.fit , datos.test)$class


#Accuracy and error

ClassAccuracy(Direction.test, qda.class)

(qda.conf <- table(Direction.test, qda.class,
                   dnn=c("Real Class", "Prediction")))

#Accuracy on test data

#Class accuracy

(accuracy.class.qda<-diag(prop.table(qda.conf,1)))

#Global accuracy

(accuracy.global.qda<-sum(diag(prop.table(qda.conf))))

mean(qda.class == datos.test$Direction)

#Error

#Error by class

(error.class.qda<-diag(1-prop.table(qda.conf,1)))

#Global error

(error.global.qda<-1-accuracy.global.qda)

1-mean(qda.class == datos.test$Direction)

#Accuracy and error with caret package

confusionMatrix(qda.class ,Direction.test)

#Variable selection

#By 10-folds cross-validation rate of correctness

(qda.step <- stepclass(Direction ~ ., "qda", direction = "both", data = datos))

qda.step$formula #best predictors

(qda.best <- qda(qda.step$formula , data = datos , subset = train))

#By Wilk's Lambda

wilks.selection <- greedy.wilks(Direction ~ ., data=datos, niveau=0.1)

wilks.selection$formula #best predictors

(qda.best.wilks <- qda(wilks.selection$formula , data = datos , subset = train))

##### Naive Bayes ####

#Fitting the model

(nb.fit <- naiveBayes(Direction ~ Lag1 + Lag2 , data = datos ,subset = train))

nb.fit #Contains mean and sd for each variable in each class

mean(datos$Lag1[train][datos$Direction[train] == "Down"])

sd(datos$Lag1[train][datos$Direction[train] == "Down"])

#Graphs

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2, method = "naiveBayes")

partimat(as.factor(datos$Direction) ~ datos$Lag1 + datos$Lag2,
         method = "naiveBayes", plot.matrix = TRUE, imageplot = FALSE)

#Prediction

nb.class <- predict(nb.fit , datos.test)

predict(nb.fit , datos.test, type="raw") #P(Y=k|X=x)

#Accuracy and error

ClassAccuracy(Direction.test, nb.class)

(nb.conf <- table(Direction.test, nb.class,
                   dnn=c("Real Class", "Prediction")))

#Accuracy on test data

#Class accuracy

(accuracy.class.nb<-diag(prop.table(nb.conf,1)))

#Global accuracy

(accuracy.global.nb<-sum(diag(prop.table(nb.conf))))

mean(nb.class == datos.test$Direction)

#Error

#Error by class

(error.class.nb<-diag(1-prop.table(nb.conf,1)))

#Global error

(error.global.nb<-1-accuracy.global.nb)

1-mean(nb.class == datos.test$Direction)

#Accuracy and error with caret package

confusionMatrix(nb.class ,Direction.test)

#Variable selection

#By Wilk's Lambda

wilks.selection <- greedy.wilks(Direction ~ ., data=datos, niveau=0.1)

wilks.selection$formula #best predictors

(nb.best.wilks <- naiveBayes(wilks.selection$formula ,
                             data = datos , subset = train))

#### K Nearest Neighbors KNN ####

#Arguments for the function

train.X <- cbind(datos$Lag1 , datos$Lag2)[train , ] #train data

test.X <- cbind(datos$Lag1 , datos$Lag2)[!train , ] #test data

train.class <- datos$Direction[train] #class on train set

#test.X = train.X to perform prediction of train set

####Prediction with k = 1####

knn1.pred <- knn(train.X, test.X, train.class , k = 1)

#Accuracy and error

ClassAccuracy(Direction.test, knn1.pred)

(knn1.conf <- table(Direction.test, knn1.pred,
                  dnn=c("Real Class", "Prediction")))

#Accuracy on test data

#Class accuracy

(accuracy.class.knn1<-diag(prop.table(knn1.conf,1)))

#Global accuracy

(accuracy.global.knn1<-sum(diag(prop.table(knn1.conf))))

mean(knn1.pred == datos.test$Direction)

#Error

#Error by class

(error.class.knn1<-diag(1-prop.table(knn1.conf,1)))

#Global error

(error.global.knn1<-1-accuracy.global.knn1)

1-mean(knn1.pred == datos.test$Direction)

#Accuracy and error with caret package

confusionMatrix(knn1.pred ,Direction.test)

####Prediction with k = sqrt(n)####

#Best k is often k = sqrt(n)

knn.sqrt.pred <- knn(train.X, test.X, train.class ,
                 k = round(sqrt(dim(datos)[1]),0))

#Accuracy and error

ClassAccuracy(Direction.test, knn.sqrt.pred)

(knn.sqrt.conf <- table(Direction.test, knn.sqrt.pred,
                    dnn=c("Real Class", "Prediction")))

#Accuracy on test data

#Class accuracy

(accuracy.class.knn.sqrt<-diag(prop.table(knn.sqrt.conf,1)))

#Global accuracy

(accuracy.global.knn.sqrt<-sum(diag(prop.table(knn.sqrt.conf))))

mean(knn.sqrt.pred == datos.test$Direction)

#Error

#Error by class

(error.class.knn.sqrt<-diag(1-prop.table(knn.sqrt.conf,1)))

#Global error

(error.global.knn.sqrt<-1-accuracy.global.knn.sqrt)

1-mean(knn.sqrt.pred == datos.test$Direction)

####Find optimal k####

#Too inefficient

i = 1

k.optm = 1

for(i in seq_along(datos[,1])){
  
  knn.mod <- knn(train=train.X, test=test.X, train.class, k=i)
  
  k.optm[i] <- 100 * sum(train.class == knn.mod)/NROW(train.class)
  
  k = i
  
  k.values = seq(1,length(k.optm))
  
  k.accuracy = tibble('k'=k.values,'Accuracy'=k.optm)
  
}

k.accuracy

plot(k.optm, type="b", xlab="K- Value",ylab="Accuracy level")

which.max(k.optm)

#Prediction with best k

knn(train=train.X, test=test.X, train.class, k=which.max(k.optm))

#Accuracy

ClassAccuracy(Direction.test, 
              knn(train=train.X, test=test.X, train.class, k=which.max(k.optm)))

#Scale of the data

#To use KNN is good to have scaled data

knn.scaled <- knn(scale(train.X), scale(test.X), train.class ,
                  k = round(sqrt(dim(datos)[1]),0))

#Comparison between methods

#Without scaling

ClassAccuracy(Direction.test,knn.sqrt.pred)

#scaled data

ClassAccuracy(Direction.test,knn.scaled)

#The best

ClassAccuracy(Direction.test,knn.sqrt.pred)$'Global Accuracy'

ClassAccuracy(Direction.test,knn.scaled)$'Global Accuracy'

#### Hypothesis Contrast #####

#### Normality ####

#Ho = Normal distributed

#H1 = Otherwise

#Density without class

density(datos$Lag1) #density function estimation

plot(density(datos$Lag1))

#Probabilistic distribution by class 

datos = tibble(datos)

datos %>% filter(Direction=='Up') %>% 
  dplyr::select(Lag1) %>% #Selection of variable
  pull() %>% #Convert tibble into vector
  density() %>% plot()

#Graph for every variable

datos.for = datos %>% dplyr::select(Year:Today)

par(mfrow=c(3,3))

for(i in seq(1,dim(datos.for)[2])){
  nom = names(datos.for)
  datos.for[,i] %>% pull() %>% density() %>% plot(main=nom[i])
}

#Everything seems to be normal distributed
 
#Graph for every variable by class

#Class Up

datos.for.up = datos %>% filter(Direction=='Up') %>% dplyr::select(Year:Today)

par(mfrow=c(3,3))

for(i in seq(1,dim(datos.for.up)[2])){
  nom = names(datos.for.up)
  datos.for.up[,i] %>% pull() %>% density() %>% plot(main=nom[i])
}

#Class Down

datos.for.down = datos %>% filter(Direction=='Down') %>% dplyr::select(Year:Today)

par(mfrow=c(3,3))

for(i in seq(1,dim(datos.for.down)[2])){
  nom = names(datos.for.down)
  datos.for.down[,i] %>% pull() %>% density() %>% plot(main=nom[i])
}

#Saphiro-Wilks

# n <=50 Test Shapiro-Wilks

#Filtering the class

datos.up = datos %>% filter(Direction=='Up') %>% dplyr::select((-logistic.pred))

shapiro.test(datos.up$Lag1) 

#For every class with taplly()

datos = as.data.frame(datos)

tapply(datos$Lag1, datos$Direction, shapiro.test)

#For every variable

datos.num = datos[,c(1:8)]

Saphiros = rep(0,dim(datos.num)[2]) #empty list

for(i in seq(1,dim(datos.num)[2])){
  
  Saphiros[i] = list( tapply(datos.num[,i], datos$Direction, shapiro.test))
  
}

nom = names(datos.num)

names(Saphiros) = nom

Saphiros

#qqnorm

qqnorm(datos.up$Lag1)
qqline(datos.up$Lag1)

#Kolmogorov-Smirnov

# n >50  #Test Kolmogorov-Smirnov #Lilliefors

lillie.test(datos.up$Lag1)

#For every variable

datos.num = datos[,c(1:8)]

Lillies = rep(0,dim(datos.num)[2]) #empty list

for(i in seq(1,dim(datos.num)[2])){
  
  Lillies[i] = list( tapply(datos.num[,i], datos$Direction, lillie.test))
  
}

nom = names(datos.num)

names(Lillies) = nom

Lillies

#Pearson

pearson.test(datos.up$Lag1)

#For every variable

datos.num = datos[,c(1:8)]

Pearsons = rep(0,dim(datos.num)[2]) #empty list

for(i in seq(1,dim(datos.num)[2])){
  
  Pearsons[i] = list( tapply(datos.num[,i], datos$Direction, pearson.test))
  
}

nom = names(datos.num)

names(Pearsons) = nom

Pearsons

#Multivariate and univariate normality test

#Without class

mvn(data = datos.num, mvnTest = "mardia",
    univariateTest = "SW",univariatePlot = "qqplot")

#By classes

datos.up = datos %>% 
  filter(Direction=='Up') %>%
  dplyr::select(-c(logistic.pred,Direction))

datos.down = datos %>% 
  filter(Direction=='Down') %>%
  dplyr::select(-c(logistic.pred,Direction))

mvn(data = datos.up, mvnTest = "mardia",
    univariateTest = "SW",univariatePlot = "qqplot")

mvn(data = datos.down, mvnTest = "mardia",
    univariateTest = "SW",univariatePlot = "qqplot")

#### Homogeneity of covariance matrix ####

boxM(datos.num, datos$Direction)

File: ClassificationMethods.py
#### Classification Methods Python ####

#### Libraries ####

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.pyplot
import scipy as scp
import sklearn
import statsmodels.api as sm
import os
import seaborn as sns
import pylab 
import scipy.stats as stats

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn import metrics 
#from sklearn.datasets import load_iris
from sklearn.metrics import (confusion_matrix,accuracy_score)
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis 
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from scipy.stats import shapiro
from scipy.stats import kstest
from scipy.stats import normaltest
from scipy.stats import levene

#Set working directory

print("Current Working Directory " , os.getcwd()) #current

try:
    # Change the current working Directory    
    os.chdir('c:/Users/98pjo/OneDrive/Escritorio')
    print("Directory changed")
except OSError:
    print("Can't change the Current Working Directory") 

#### Data ####

#Import the data from .txt

datos = pd.read_table('flores.txt', sep=" ", header='infer')

#iris_data = load_iris()
#datos = pd.DataFrame(iris_data.data, columns=iris_data.feature_names)
#Other way to get the same data

#Convert into categorical variable 

datos['Species'] = datos['Species'].astype('category')

#Make a categorical variable with nums

cat_columns = datos.select_dtypes(['category']).columns

datos['code'] = datos[cat_columns].apply(lambda x: x.cat.codes)

#Define Y as response variable

Y = datos['Species']

#Define X as the predictors 

X = datos.drop(['Species','code'], axis=1)

#Statistical summary
datos.describe() 

#Formula

formula = 'Species~Sepal.Length+Sepal.Width+Petal.Length+Petal.Width'

#Correlation matrix

datos.corr()

#Graphs by Specie

label = datos['code'] #need factor as numerical

colors = ['red','green','blue']

fig = plt.figure(figsize=(8,8))

x = np.array(datos['Petal.Length'])

y = np.array(datos['Petal.Width'])

plt.scatter(x, y, c=label, cmap=matplotlib.colors.ListedColormap(colors))

del(x,y) #delete instrumental variables for the plot

#Split data into train and test subsets

#Using pandas sample #70% train

training_data = datos.sample(frac=0.7, random_state=25)

testing_data = datos.drop(training_data.index)

print(f"No. of training examples: {training_data.shape[0]}")

print(f"No. of testing examples: {testing_data.shape[0]}")

#Using scikit-learn #Test 30%

training_data, testing_data = train_test_split(datos, test_size=0.3,
                                               random_state=25)

#Train and test data for Y and X variables

X_train, X_test, Y_train, Y_test = train_test_split(X, Y,
                                                    test_size = 0.30,
                                                    random_state = 5)

#### Logistic Regression ####

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

logistic_fit = LogisticRegression(random_state=0,
                   multi_class='multinomial',
                   penalty='none', solver='newton-cg').fit(X_train, Y_train)

#Parameters

params = logistic_fit.get_params()

print(params)

#Print model parameters

# 1 equation by Y class
# B gives the relation between logit and X

print('Intercept: \n', logistic_fit.intercept_)

print('Variables: \n', X.columns)
print('Coefficients: \n', logistic_fit.coef_)

logistic_fit_coefficients = pd.concat([pd.DataFrame(X_train.columns),
                          pd.DataFrame(np.transpose(logistic_fit.coef_))],
                         axis = 1)

print(logistic_fit_coefficients)

#Prediction on test data

logistic_pred = logistic_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, logistic_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, logistic_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, logistic_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, logistic_pred))

#Accuracy by class

print(classification_report(Y_test, logistic_pred, digits=4))

#### Linear Discriminant Analysis ####

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

lda_fit = LinearDiscriminantAnalysis().fit(X_train, Y_train)

#Parameters

params = lda_fit.get_params()

print(params)

#Print model parameters

# 1 equation by Y class
# B gives the relation between logit and X

print('Intercept: \n', lda_fit.intercept_)

print('Variables: \n', X.columns)
print('Coefficients: \n', lda_fit.coef_)

lda_fit_coefficients = pd.concat([pd.DataFrame(X_train.columns),
                          pd.DataFrame(np.transpose(lda_fit.coef_))],
                         axis = 1)

print(lda_fit_coefficients)

#Prediction on test data

lda_pred = lda_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, lda_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, lda_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, lda_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, lda_pred))

#Accuracy by class

print(classification_report(Y_test, lda_pred, digits=4))

#Graph

#define data to plot
x = np.array(X)
y = np.array(datos['code'])
model = LinearDiscriminantAnalysis()
data_plot = model.fit(x, y).transform(x)
target_names = ['setosa','versicolor','virginica']

#create LDA plot
plt.figure()
colors = ['red', 'green', 'blue']
lw = 2
for color, i, target_name in zip(colors, [0, 1, 2], target_names):
    plt.scatter(data_plot[y == i, 0], data_plot[y == i, 1], alpha=.8, color=color,
                label=target_name)

#add legend to plot
plt.legend(loc='best', shadow=False, scatterpoints=1)

#display LDA plot
plt.show()

### Quadratic Discriminant Analysis ###

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

qda_fit = QuadraticDiscriminantAnalysis().fit(X_train, Y_train)

#Prediction on test data

qda_pred = qda_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, qda_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, qda_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, qda_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, qda_pred))

#Accuracy by class

print(classification_report(Y_test, qda_pred, digits=4))

### Naive Bayes Classifier###

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

nb_fit = GaussianNB().fit(X_train, Y_train)

#Prediction on test data

nb_pred = nb_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, nb_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, nb_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, nb_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, nb_pred))

#Accuracy by class

print(classification_report(Y_test, nb_pred, digits=4))

### K-Nearest Neighbors k = 1 ###

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

knn1_fit = KNeighborsClassifier(n_neighbors=1).fit(X_train, Y_train)

#Prediction on test data

knn1_pred = knn1_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, knn1_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, knn1_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, knn1_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, knn1_pred))

#Accuracy by class

print(classification_report(Y_test, knn1_pred, digits=4))

### K-Nearest Neighbors k = sqrt(n) ###

## With sklearn
#X_train must contain the X we want to use

#Fitting the model on train data

n = int(round(np.sqrt(len(datos)),0))

knn_sqrt_fit = KNeighborsClassifier(n_neighbors=n).fit(X_train, Y_train)

#Prediction on test data

knn_sqrt_pred = knn_sqrt_fit.predict(X_test)

#Accuracy

#Confusion Matrix

confusion_matrix(Y_test, knn_sqrt_pred)

print ("Confusion Matrix : \n", confusion_matrix(Y_test, knn_sqrt_pred))

#Accuracy score of the model

print('Test accuracy = ', accuracy_score(Y_test, knn_sqrt_pred))

#Error of the model

print('Test error = ', 1-accuracy_score(Y_test, knn_sqrt_pred))

#Accuracy by class

print(classification_report(Y_test, knn_sqrt_pred, digits=4))

### Hypothesis Contrast ###

## Normality

#Ho = Normal distributed

#H1 = Otherwise

#Histogram and density 

# matplotlib histogram

plt.hist(datos['Sepal.Length'], color = 'blue', edgecolor = 'black',
         bins = int(180/5))

# seaborn histogram

sns.distplot(datos['Sepal.Length'], hist=True, kde=False, 
             bins=int(180/5), color = 'blue',
             hist_kws={'edgecolor':'black'})

# Density Plot and Histogram of all arrival delays

sns.distplot(datos['Sepal.Length'], hist=True, kde=True, 
             bins=int(180/5), color = 'darkblue', 
             hist_kws={'edgecolor':'black'},
             kde_kws={'linewidth': 1})

# Every variable

variables = X.columns

# Iterate through the variables
for variables in variables:
    sns.distplot(datos[variables], hist=False, kde=True, 
                 bins=int(180/5), color = 'darkblue', 
                 hist_kws={'edgecolor':'black'},
                 kde_kws={'linewidth': 1})


# Iterate through the five airlines
for variables in variables:
    
    # Draw the density plot
    sns.distplot(datos[variables], hist = False, kde = True,
                 kde_kws = {'linewidth': 1},
                 label = variables)

#Saphiro-Wilks

# n <=50 Test Shapiro-Wilks

shapiro(datos['Sepal.Length'])

#For every variable

X_num = np.array(X)

for i in range(0,X_num.shape[1]):
    print(shapiro(X_num[:,i]))

#By class

#Select class

datos_class1 = datos[datos.Species=='virginica']

X_class1 = np.array(datos_class1.drop(['Species','code'], axis=1))

#Test for every variable

for i in range(0,X_class1.shape[1]):
    print(shapiro(X_class1[:,i]))

#qqnorm

stats.probplot(datos['Sepal.Length'], dist="norm", plot=pylab)

pylab.show()

#Kolmogorov-Smirnov

# n >50  #Test Kolmogorov-Smirnov #Lilliefors

# n <=50 Test Shapiro-Wilks

kstest(datos['Sepal.Length'],'norm')

#For every variable

X_num = np.array(X)

for i in range(0,X_num.shape[1]):
    print(kstest(X_num[:,i],'norm'))

#By class

#Select class

datos_class1 = datos[datos.Species=='virginica']

X_class1 = np.array(datos_class1.drop(['Species','code'], axis=1))

#Test for every variable

for i in range(0,X_class1.shape[1]):
    print(kstest(X_class1[:,i],'norm'))

#Pearson

normaltest(datos['Sepal.Length'])

#For every variable

X_num = np.array(X)

for i in range(0,X_num.shape[1]):
    print(normaltest(X_num[:,i]))

#By class

#Select class

datos_class1 = datos[datos.Species=='virginica']

X_class1 = np.array(datos_class1.drop(['Species','code'], axis=1))

#Test for every variable

for i in range(0,X_class1.shape[1]):
    print(normaltest(X_class1[:,i]))

## Homogeneity of covariance matrix

#Levene's test

datos_class1 = datos[datos.Species=='virginica']

datos_class2 = datos[datos.Species=='setosa']

datos_class3 = datos[datos.Species=='versicolor']

levene(datos_class1['Sepal.Length'],
       datos_class2['Sepal.Length'],
       datos_class3['Sepal.Length'],
       center='median')

#For every variable

for i in range(0,X_num.shape[1]):
    print(levene(np.array(datos_class1)[:,i],
           np.array(datos_class2)[:,i],
           np.array(datos_class3)[:,i],
           center='median'))













